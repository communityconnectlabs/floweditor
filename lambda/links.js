const result = {
  id: 1,
  name: 'Links',
  links: [
    {
      node_uuid: '6f553e7b-a52d-4638-8b19-9e6e5ebed4dd',
      action_uuid: 'd8ddd46f-8e96-46bf-a46f-626a3d275dd6',
      link: 'https://google.com',
      status_code: 200
    },
    {
      node_uuid: '6f553e7b-a52d-4638-8b19-9e6e5ebed4dd',
      action_uuid: 'd8ddd46f-8e96-46bf-a46f-626a3d275dd6',
      link: 'https://twitter.com',
      status_code: 200
    },
    {
      node_uuid: '858f78c9-ff10-4e19-83ab-10762320f8db',
      action_uuid: '49e2047b-9dc6-4ae9-b9be-6607488357f4',
      link: 'https://hauyfjwgfjw.com/',
      status_code: 400,
      error: 'A Connection error occurred.'
    }
  ]
};
const { getOpts } = require('./utils');

exports.handler = (evt, ctx, cb) => cb(null, getOpts({ body: JSON.stringify(result) }));
