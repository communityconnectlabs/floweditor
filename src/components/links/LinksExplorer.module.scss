@import 'variables.module.scss';

.message {
  padding: 12px 12px;

  .header {
    display: inline-block;
  }

  &:hover {
    background: $light_gray_2;
    cursor: pointer;
  }
}

.links_wrapper {
  background-color: #fff;
  width: 250px;
  height: 325px;
  overflow-y: auto;
  overflow-x: hidden;
  display: flex;
  flex-direction: column;
  justify-items: start;

  .links {
    background: #fff;
    height: 325px;
    overflow-y: auto;
    overflow-x: hidden;

    .link {
      color: $text_color;
      padding: 11px 10px;
      font-size: 11px;
      display: flex;

      &:hover {
        background: $light_gray_2;
        cursor: pointer;
      }

      .status_code {
        margin-left: auto;
        font-size: 14px;
        font-weight: bold;
        text-decoration: none;
        color: $text_color;

        &.success {
          color: $link_color;
          &:hover {
            text-shadow: $link_color 1px 0 10px;
          }
        }

        &.error {
          color: $error_color;
          &:hover {
            text-shadow: $error_color 1px 0 10px;
          }
        }
      }
    }
  }

  .buttons {
    margin-top: auto;
    display: flex;
    flex-direction: column;
    justify-self: end;
    padding: 5px;
    gap: 5px;

    .last_updated {

    }

    .validate {
      text-align: center;
      display: inline-block;
      vertical-align: top;
      font-weight: 400;
      font-size: 13px;
      cursor: pointer;
      border-radius: 4px;
      background-image: linear-gradient(to bottom, #289f9b, #289f9b);
      color: #ffffff;
      padding: 8px 16px 8px 16px;
      text-decoration: none;
      -webkit-user-select: none;
      user-select: none;
      border: 1px solid rgba(90, 90, 90, 0.1);

      &:hover {
        background-image: linear-gradient(to bottom, #458c8a, #458c8a);
        text-decoration: none;
      }
    }

    .loading {
      display: flex;
      justify-content: center;
    }
  }
}

.language {
  padding: 6px 10px;
  color: #fff;
  font-weight: 400;
  background: $red;
}

.mask {
  position: fixed;
  left: 0;
  top: 0;
  background: rgba(0, 0, 0, 0.5);
  height: 100%;
  width: 100%;
  opacity: 0;
  display: none;
  z-index: -1;
  transition: opacity 0.6s;
}

.visible {
  .mask {
    opacity: 1;
    // display: block;
    z-index: $z_pop_tabs + 1;
  }
}

.hidden {
  visibility: hidden;
  z-index: 0;
}