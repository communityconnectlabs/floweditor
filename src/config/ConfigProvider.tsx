/* eslint-disable @typescript-eslint/no-explicit-any */
import { FlowEditorConfig } from 'flowTypes';
import * as React from 'react';

export const fakePropType: any = (): any => null;
fakePropType.isRequired = (): any => null;

interface ConfigProviderProps {
  config: FlowEditorConfig;
  children: any;
}

export interface ConfigProviderContext {
  config: FlowEditorConfig;
}

// ----------------------------------------------------------------------------------------------

export const SINGLE_CHILD_ERROR = 'ConfigProvider expects only one child component.';
export const VALID_CHILD_ERROR =
  'ConfigProvider expects a valid React element: https://reactjs.org/docs/react-api.html#isvalidelement';

export default class ConfigProvider extends React.Component<ConfigProviderProps> {
  public static childContextTypes = {
    config: fakePropType,
    assetService: fakePropType
  };

  constructor(props: ConfigProviderProps) {
    super(props);
    const win = window as any;
    if (win.isMobile && win.isMobile()) {
      this.props.config.mutable = false;
    }

    if (React.Children.count(props.children) > 1) {
      throw new Error(SINGLE_CHILD_ERROR);
    } else if (!React.isValidElement(props.children)) {
      throw new Error(VALID_CHILD_ERROR);
    }
  }

  public getChildContext(): ConfigProviderContext {
    return {
      config: this.props.config
    };
  }

  public render(): JSX.Element {
    return React.Children.only(this.props.children);
  }
}
