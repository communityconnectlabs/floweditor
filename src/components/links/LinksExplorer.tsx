import { react as bindCallbacks } from 'auto-bind';
import classNames from 'classnames/bind';
import { PopTab } from 'components/poptab/PopTab';
import { FlowDefinition, FlowDetails } from 'flowTypes';
import React from 'react';

import styles from './LinksExplorer.module.scss';
import i18n from 'config/i18n';
import { PopTabType } from 'config/interfaces';
import { AssetStore } from '../../store/flowContext';
import { renderIf } from '../../utils';
import Loading from '../loading/Loading';
import dateFormat from 'dateformat';

const cx: any = classNames.bind(styles);

export interface User {
  email: string;
  name: string;
}

export interface LinkData {
  node_uuid: string;
  action_uuid: string;
  link: string;
  error?: string;
  status_code?: number;
  processing?: boolean;
}

export interface LinksExplorerProps {
  utc?: boolean;
  popped: string;
  linksUrl: string;
  position: string;
  onToggled: (visible: boolean, tab: PopTabType) => void;
  onLinkClicked: (node_uuid: string, action_uuid: string) => void;
  loadFlowDefinition: (details: FlowDetails, assetStore: AssetStore) => void;
}

export interface LinksExplorerState {
  definition: FlowDefinition;
  visible: boolean;
  links: Array<LinkData>;
  lastRun?: string | Date;
  processing?: boolean;
}

export class LinksExplorer extends React.Component<LinksExplorerProps, LinksExplorerState> {
  constructor(props: LinksExplorerProps) {
    super(props);

    this.state = {
      definition: null,
      visible: false,
      links: [],
      processing: true,
      lastRun: new Date()
    };

    bindCallbacks(this, {
      include: [/^handle/, /^render/]
    });
  }

  public componentDidMount() {
    this.loadValidatedLinks(false);
  }

  public handleTabClicked(): void {
    this.props.onToggled(!this.state.visible, PopTabType.LINKS_TAB);

    this.setState(
      (prevState: LinksExplorerState) => {
        return { visible: !prevState.visible };
      },
      () => {}
    );
  }

  private loadValidatedLinks(refresh: boolean) {
    let linksUrl = this.props.linksUrl;
    if (refresh) {
      linksUrl += '&refresh=true';
      this.setState({ links: [], processing: true });
    }
    fetch(linksUrl)
      .then(response => response.json())
      .then(data => {
        if (data.validating) {
          setTimeout(this.loadValidatedLinks.bind(this), data.refresh_timeout);
        } else {
          this.setState({
            links: data.links,
            lastRun: data.processed_on,
            processing: false
          });
        }
      })
      .catch(() => {
        setTimeout(this.loadValidatedLinks.bind(this), 10000);
      });
  }

  private onRefreshClicked() {
    this.loadValidatedLinks(true);
  }

  public render(): JSX.Element {
    const classes = cx({
      [styles.visible]: this.state.visible,
      [styles.hidden]: this.props.popped && this.props.popped !== PopTabType.LINKS_TAB
    });

    return (
      <div className={classes}>
        <div className={styles.mask} />
        <PopTab
          header={i18n.t('links.header', 'Links')}
          color="#289f9b"
          icon="fe-link"
          label={i18n.t('links.label', 'Validate Links')}
          top={this.props.position}
          popTop={'-60px'}
          visible={this.state.visible}
          onShow={this.handleTabClicked}
          onHide={this.handleTabClicked}
          custom={false}
        >
          <div className={styles.links_wrapper}>
            <div className={styles.links}>
              {this.state.links.map((link, key) => {
                return (
                  <div
                    className={styles.link}
                    key={key}
                    onClick={() => {
                      this.props.onLinkClicked(link.node_uuid, link.action_uuid);
                    }}
                  >
                    {link.link}
                    {renderIf(!!link.status_code)(
                      <a
                        href={link.link}
                        target="_blank"
                        rel="noopener noreferrer"
                        className={
                          styles.status_code +
                          ' ' +
                          (link.status_code >= 400 ? styles.error : styles.success)
                        }
                      >
                        {link.status_code}
                      </a>
                    )}
                  </div>
                );
              })}
            </div>
            <div className={styles.buttons}>
              {renderIf(!this.state.processing)(
                <>
                  <div className={styles.last_updated}>
                    Last run:{' '}
                    {dateFormat(this.state.lastRun, 'mmmm d, yyyy, h:MM TT', this.props.utc)}
                  </div>
                  <button className={styles.validate} onClick={this.onRefreshClicked.bind(this)}>
                    {'Validate Links'}
                  </button>
                </>
              )}
              {renderIf(this.state.processing)(
                <div className={styles.loading}>
                  <Loading size={10} units={5} color={'#289f9b'} />
                </div>
              )}
            </div>
          </div>
        </PopTab>
      </div>
    );
  }
}
