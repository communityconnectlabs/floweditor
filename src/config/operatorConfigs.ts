import { HIDDEN, Operator, OperatorMap, Operators, TEXT_TYPES, ONLINE } from 'config/interfaces';

export const intentOperatorList: Operator[] = [
  {
    type: Operators.has_top_intent,
    verboseName: 'has top intent',
    operands: 2,
    visibility: ONLINE
  },
  {
    type: Operators.has_intent,
    verboseName: 'has intent',
    operands: 2,
    visibility: ONLINE
  }
];

export const operatorConfigList: Operator[] = [
  {
    type: Operators.has_any_word,
    verboseName: 'has any of the words',
    operands: 1,
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_all_words,
    verboseName: 'has all of the words',
    operands: 1,
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_phrase,
    verboseName: 'has the phrase',
    operands: 1,
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_only_phrase,
    verboseName: 'has only the phrase',
    operands: 1,
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_beginning,
    verboseName: 'starts with',
    operands: 1
  },
  {
    type: Operators.has_text,
    verboseName: 'has some text',
    operands: 0,
    categoryName: 'Has Text',
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_number,
    verboseName: 'has a number',
    operands: 0,
    categoryName: 'Has Number'
  },
  {
    type: Operators.has_number_between,
    verboseName: 'has a number between',
    operands: 2
  },
  {
    type: Operators.has_number_lt,
    verboseName: 'has a number below',
    operands: 1
  },
  {
    type: Operators.has_number_lte,
    verboseName: 'has a number at or below',
    operands: 1
  },
  {
    type: Operators.has_number_eq,
    verboseName: 'has a number equal to',
    operands: 1
  },
  {
    type: Operators.has_number_gte,
    verboseName: 'has a number at or above',
    operands: 1
  },
  {
    type: Operators.has_number_gt,
    verboseName: 'has a number above',
    operands: 1
  },
  {
    type: Operators.has_date,
    verboseName: 'has a date',
    operands: 0,
    categoryName: 'Has Date',
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_date_lt,
    verboseName: 'has a date before',
    operands: 1,
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_date_eq,
    verboseName: 'has a date equal to',
    operands: 1,
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_date_gt,
    verboseName: 'has a date after',
    operands: 1,
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_time,
    verboseName: 'has a time',
    operands: 0,
    categoryName: 'Has Time',
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_group,
    verboseName: 'is in the group',
    operands: 1,
    visibility: HIDDEN
  },
  {
    type: Operators.has_category,
    verboseName: 'has the category',
    operands: 0,
    visibility: HIDDEN
  },
  {
    type: Operators.has_phone,
    verboseName: 'has a phone number',
    operands: 0,
    categoryName: 'Has Phone'
  },
  {
    type: Operators.has_email,
    verboseName: 'has an email',
    operands: 0,
    categoryName: 'Has Email',
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_state,
    verboseName: 'has state',
    operands: 0,
    categoryName: 'Has State'
  },
  {
    type: Operators.has_district,
    verboseName: 'has district',
    operands: 1,
    categoryName: 'Has District'
  },
  {
    type: Operators.has_ward,
    verboseName: 'has ward',
    operands: 2,
    categoryName: 'Has Ward'
  },
  {
    type: Operators.has_error,
    verboseName: 'has an error',
    operands: 0,
    categoryName: 'Has Error',
    visibility: HIDDEN
  },
  {
    type: Operators.has_value,
    verboseName: 'is not empty',
    operands: 0,
    categoryName: 'Not Empty'
  },
  {
    type: Operators.has_pattern,
    verboseName: 'matches regex',
    operands: 1
  }
];

export const operatorConfigMap: OperatorMap = [...operatorConfigList, ...intentOperatorList].reduce(
  (map: OperatorMap, operatorConfig: Operator) => {
    map[operatorConfig.type] = operatorConfig;
    return map;
  },
  {}
);

/**
 * Shortcut for constant lookup of operator config in operator configs map
 * @param {string} type - The type of the operator config to return, e.g. 'send_msg'
 * @returns {Object} - The operator config found at operatorConfigs[type] or -1
 */
export const getOperatorConfig = (type: Operators): Operator => operatorConfigMap[type];
