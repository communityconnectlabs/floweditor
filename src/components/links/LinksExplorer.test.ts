import { composeComponentTestUtils } from 'testUtils';
import { LinksExplorer, LinksExplorerProps } from 'components/links/LinksExplorer';
import { PopTabType } from '../../config/interfaces';

const baseProps: LinksExplorerProps = {
  onToggled: jest.fn(),
  popped: PopTabType.LINKS_TAB
};

const { setup } = composeComponentTestUtils(LinksExplorer, baseProps);

describe(LinksExplorer.name, () => {
  describe('render', () => {
    it('should render base component', async () => {
      const { wrapper, instance } = setup();
      expect(wrapper).toMatchSnapshot();
    });
  });
});
