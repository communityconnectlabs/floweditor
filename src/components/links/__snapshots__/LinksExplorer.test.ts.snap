// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinksExplorer render should render base component 1`] = `
<div
  className=""
>
  <div
    className="mask"
  />
  <PopTab
    color="#289f9b"
    custom={false}
    header="Links"
    icon="fe-link"
    label="Validate Links"
    onHide={[Function]}
    onShow={[Function]}
    popTop="-60px"
    visible={false}
  >
    <div
      className="links_wrapper"
    >
      <div
        className="links"
      />
      <div
        className="buttons"
      >
        <div
          className="loading"
        >
          <Loading
            color="#289f9b"
            size={10}
            units={5}
          />
        </div>
      </div>
    </div>
  </PopTab>
</div>
`;
